File: solutions/018/018.c
#include <stdio.h>
#include <stdlib.h>

#define NO_OF_ROWS 15


int maximumPathSum(int tree[NO_OF_ROWS][NO_OF_ROWS]) {
    for (int i = NO_OF_ROWS - 2; i >= 0; i--) {
        for (int j = 0; j <= i; j++) {
            int x = tree[i+1][j];
            int y = tree[i+1][j+1];
            tree[i][j] += (x > y) ? x : y;
        }
    }
    return tree[0][0];
}


int main() {
    FILE *fp = fopen("./resources/tree.txt", "r");

    int tree[NO_OF_ROWS][NO_OF_ROWS];

    int row = 0;
    int col = 0;

    char tmp[3];
    char *ptr = tmp;

    int c;

    while ((c = fgetc(fp)) != EOF) {
        if (c == ' ') {
            tree[row][col++] = atoi(tmp);
            ptr = tmp;
        } else if (c == '\n') {
            tree[row][col++] = atoi(tmp);
            ptr = tmp;
            row++;
            col = 0;
        } else {
            *ptr++ = c;
        }
    }

    printf("%d\n", maximumPathSum(tree));
}

File: solutions/018/018.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func maximumPathSum(tree [][]int) int {

	for i := len(tree) - 2; i >= 0; i-- {
		row := tree[i]

		for j := 0; j < len(row); j++ {
			x := tree[i+1][j]
			y := tree[i+1][j+1]

			if x > y {
				row[j] += x
			} else {
				row[j] += y
			}
		}
	}

	return tree[0][0]
}

func main() {
	bs, _ := ioutil.ReadFile("./resources/tree.txt")
	lines := strings.Split(string(bs), "\n")

	tree := [][]int{}

	// Remove empty line at end
	lines = lines[0 : len(lines)-1]

	for _, line := range lines {
		row := []int{}
		numbers := strings.Split(line, " ")
		for _, number := range numbers {
			n, _ := strconv.Atoi(number)
			row = append(row, n)
		}
		tree = append(tree, row)
	}

	fmt.Println(maximumPathSum(tree))
}
